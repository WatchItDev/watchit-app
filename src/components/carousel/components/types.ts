import {BoxProps} from "@mui/material/Box";
import {SxProps, Theme} from "@mui/material/styles";
import {IconifyProps} from "@src/components/iconify";
import {StackProps} from "@mui/material/Stack";
import React from "react";
import {IconButtonProps} from "@mui/material/IconButton";

export interface CarouselDotsProps extends BoxProps {
  rounded?: boolean;
  sx?: SxProps<Theme>;
}

export type ArrowIconProps = {
  icon?: IconifyProps; // Right icon
  isRTL?: boolean;
};

export type CarouselArrowIndexProps = {
  index: number;
  total: number;
  icon?: IconifyProps; // Right icon
  onNext?: VoidFunction;
  onPrev?: VoidFunction;
  sx?: SxProps<Theme>;
};

export interface CarouselArrowsProps extends StackProps {
  shape?: 'circular' | 'rounded';
  filled?: boolean;
  children?: React.ReactNode;
  icon?: IconifyProps; // Right icon
  onNext?: VoidFunction;
  onPrev?: VoidFunction;
  leftButtonProps?: IconButtonProps;
  rightButtonProps?: IconButtonProps;
}

export interface CarouselArrowsStyledIconButtonProps extends IconButtonProps {
  filled?: boolean;
  shape?: 'circular' | 'rounded';
  hasChild?: boolean;
}

export interface NavigationArrowsProps {
  next: () => void;
  prev: () => void;
}

export type CarouselDotsStyledRootProps = {
  rounded: boolean;
};

