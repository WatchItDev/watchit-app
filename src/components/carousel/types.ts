import React from "react";

import {Profile, VideoMetadataV3} from "@lens-protocol/api-bindings";
// @ts-ignore
import { Post,PublicationMetadataMediaVideoFieldPolicy } from "@lens-protocol/api-bindings/dist/declarations/src/lens/graphql/generated";

import {StackProps} from "@mui/material/Stack";
import {IconButtonProps} from "@mui/material/IconButton";

import {BoxProps} from "@mui/material/Box";
import {SxProps, Theme} from "@mui/material/styles";
import {IconifyProps} from "@src/components/iconify";


export interface CarouselDotsProps extends BoxProps {
  rounded?: boolean;
  sx?: SxProps<Theme>;
}

interface CarouselNavigationActions {
  onNext?: VoidFunction;
  onPrev?: VoidFunction;
}

export type ArrowIconProps = {
  icon?: IconifyProps;
  isRTL?: boolean;
};

export type CarouselArrowIndexProps = CarouselNavigationActions & {
  index: number;
  total: number;
  icon?: IconifyProps;
  sx?: SxProps<Theme>;
};

export interface CarouselArrowsProps extends StackProps, CarouselNavigationActions {
  shape?: 'circular' | 'rounded';
  filled?: boolean;
  children?: React.ReactNode;
  icon?: IconifyProps; // Right icon
  leftButtonProps?: IconButtonProps;
  rightButtonProps?: IconButtonProps;
}

export interface CarouselArrowsStyledIconButtonProps extends IconButtonProps {
  filled?: boolean;
  shape?: 'circular' | 'rounded';
  hasChild?: boolean;
}

export interface NavigationArrowsProps {
  next: () => void;
  prev: () => void;
}

export type CarouselDotsStyledRootProps = {
  rounded: boolean;
};

interface CarouselResizable {
  minItemWidth: number;
  maxItemWidth: number;
}

export interface CarouselWrapperProps<T> extends CarouselResizable{
  data: T[];
  title?: string;
  renderSlide: (slideItems: T[], itemsPerRow: number, index: number) => React.ReactNode;
  carouselSettings: any;
  boxStyle?: any;
  boxClassName?: string;
}

export interface CarouselSlideProps<T> {
  items: T[];
  itemsPerRow: number;
  renderItem: (item: T) => React.ReactNode;
}

export type CarouselCreatorsProps = CarouselResizable & {
  data: Profile[];
  title?: string;
};

export type CarouselPosterMiniProps = CarouselResizable & {
  data: Post[];
  title?: string;
};

export type CarouselTopTitlesProps = {
  posts: Post[];
  category?: string;
};

export type CarouselPosterSlideProps = {
  items: Post[];
  itemsPerRow: number;
};

export interface PublicationType extends Post {
  altTag: string;
  metadata: VideoMetadataV3 | PublicationMetadataMediaVideoFieldPolicy;
}
