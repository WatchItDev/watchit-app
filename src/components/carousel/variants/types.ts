import {Profile} from "@lens-protocol/api-bindings";
// @ts-ignore
import type {Post} from "@lens-protocol/api-bindings/dist/declarations/src/lens/graphql/generated";
import React from "react";

export interface CarouselWrapperProps<T> {
  data: T[];
  title?: string;
  minItemWidth: number;
  maxItemWidth: number;
  renderSlide: (slideItems: T[], itemsPerRow: number, index: number) => React.ReactNode;
  carouselSettings: any;
  boxStyle?: any;
  boxClassName?: string;
}

export type CarouselSlideProps = {
  items: Profile[];
  itemsPerRow: number;
};

export type CarouselCreatorsProps = {
  data: Profile[];
  title?: string;
  minItemWidth: number;
  maxItemWidth: number;
};

export type CarouselPosterMiniProps = {
  data: Post[];
  title?: string;
  minItemWidth: number;
  maxItemWidth: number;
};

export type CarouselTopTitlesProps = {
  posts: Post[];
  category?: string;
};

export type CarouselPosterSlideProps = {
  items: Post[];
  itemsPerRow: number;
};
