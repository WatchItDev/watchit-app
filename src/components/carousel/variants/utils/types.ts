export interface calculateItemsPerSlideProps {
  parentWidth: number;
  minItemWidth: number;
  maxItemWidth: number;
}
