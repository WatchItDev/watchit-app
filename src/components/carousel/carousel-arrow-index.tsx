// @mui
import { useTheme, styled, Theme, SxProps } from '@mui/material/styles';
import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
// theme
import { bgBlur } from 'src/theme/css';
//
import { IconifyProps } from '../iconify';
//
import { LeftIcon, RightIcon } from './arrow-icons';

// ----------------------------------------------------------------------

const StyledRoot = styled(Box)(({ theme }) => ({
  ...bgBlur({
    opacity: 0.48,
    color: theme.palette.grey[900],
  }),
  zIndex: 9,
  display: 'inline-flex',
  alignItems: 'center',
  position: 'absolute',
  bottom: theme.spacing(2),
  right: theme.spacing(2),
  padding: theme.spacing(0.25),
  color: theme.palette.common.white,
  borderRadius: theme.shape.borderRadius,
}));

const StyledIconButton = styled(IconButton)({
  width: 28,
  height: 28,
  padding: 0,
  opacity: 0.48,
  '&:hover': { opacity: 1 },
});

// ----------------------------------------------------------------------

type Props = {
  index: number;
  total: number;
  icon?: IconifyProps; // Right icon
  onNext?: VoidFunction;
  onPrev?: VoidFunction;
  sx?: SxProps<Theme>;
};

export default function CarouselArrowIndex({
  index,
  total,
  onNext,
  onPrev,
  icon,
  sx,
  ...other
}: Props) {
  const theme = useTheme();

  const isRTL = theme.direction === 'rtl';

  return (
    <StyledRoot sx={sx} {...other}>
      <StyledIconButton color="inherit" onClick={onPrev}>
        <LeftIcon icon={icon} isRTL={isRTL} />
      </StyledIconButton>

      <Typography variant="subtitle2" component="span" sx={{ mx: 0.25 }}>
        {index + 1}/{total}
      </Typography>

      <StyledIconButton color="inherit" onClick={onNext}>
        <RightIcon icon={icon} isRTL={isRTL} />
      </StyledIconButton>
    </StyledRoot>
  );
}
