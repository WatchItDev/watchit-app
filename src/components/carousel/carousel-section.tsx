// REACT IMPORTS
import { FC, PropsWithChildren } from 'react';

// MUI IMPORTS
import CardHeader from '@mui/material/CardHeader';
import CardContent from '@mui/material/CardContent';
import Card from '@mui/material/Card';

// ----------------------------------------------------------------------

type CarouselSectionProps = {
  title?: string
};

export const CarouselSection: FC<PropsWithChildren<CarouselSectionProps>> = ({ title, children }) => (
    <Card>
      <CardHeader title={title} sx={{ px: '50px' }} />
      <CardContent sx={{ px: '50px' }}>
        { children }
      </CardContent>
    </Card>
)
