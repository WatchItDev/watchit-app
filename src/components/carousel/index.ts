import Carousel from 'react-slick';

export { default as CarouselDots } from './carousel-dots';
export { default as CarouselArrows } from './carousel-arrows';
export { default as CarouselArrowIndex } from './carousel-arrow-index';

export { default as useCarousel } from './use-carousel';

export default Carousel;
