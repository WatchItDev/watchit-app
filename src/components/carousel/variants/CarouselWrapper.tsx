import Box from '@mui/material/Box';
import Carousel, { useCarousel } from '@src/components/carousel/index';
import CarouselNavigationArrows from '@src/components/carousel/components/CarouselNavigationArrows.tsx';
import { CarouselSection } from '@src/components/poster/carousel-section.tsx';
import { useItemsPerSlide } from '@src/hooks/components/use-item-per-slide';
import { useChunkedData } from '@src/hooks/components/use-chunked-data';
import { CarouselWrapperProps } from '../types';

export default function CarouselWrapper<T>(props: Readonly<CarouselWrapperProps<T>>) {
  const {
    data,
    title,
    minItemWidth,
    maxItemWidth,
    renderSlide,
    carouselSettings,
    boxStyle,
    boxClassName,
  } = props;

  const { itemsPerSlide, parentRef } = useItemsPerSlide({ minItemWidth, maxItemWidth });
  const slideData = useChunkedData<T>(data, itemsPerSlide);

  const carousel = useCarousel({
    ...carouselSettings,
  });

  return (
    <CarouselSection
      title={title}
      action={<CarouselNavigationArrows next={carousel.onNext} prev={carousel.onPrev} />}
    >
      <Box
        ref={parentRef}
        sx={{
          overflow: 'hidden',
          position: 'relative',
          ...boxStyle,
        }}
        className={boxClassName}
      >
        <Carousel ref={carousel.carouselRef} {...carousel.carouselSettings}>
          {slideData.map((slideItems, index) => renderSlide(slideItems, itemsPerSlide, index))}
        </Carousel>
      </Box>
    </CarouselSection>
  );
}
