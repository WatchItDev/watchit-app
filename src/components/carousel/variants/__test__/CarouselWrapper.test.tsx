import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import CarouselWrapper from '@src/components/carousel/variants/CarouselWrapper';
import { CarouselWrapperProps } from '@src/components/carousel/types';

describe('[COMPONENTS]: CarouselWrapper', () => {
  const mockData = [
    { id: '1', metadata: { title: 'Test Title 1', content: 'Test Content 1' } },
    { id: '2', metadata: { title: 'Test Title 2', content: 'Test Content 2' } },
  ];

  const defaultProps: CarouselWrapperProps<typeof mockData[0]> = {
    data: mockData,
    title: 'Test Carousel',
    minItemWidth: 100,
    maxItemWidth: 200,
    renderSlide: (slideItems, itemsPerRow, index) => (
      <div key={`slide-${index}`} data-testid="slide">
        {slideItems.map((item) => (
          <div key={item.id} data-testid="item">
            {item.metadata.title}
          </div>
        ))}
      </div>
    ),
    carouselSettings: {
      infinite: false,
      slidesToShow: 1,
      speed: 500,
      rows: 1,
      slidesPerRow: 1,
      adaptiveHeight: true,
      focusOnSelect: true,
      swipeToSlide: true,
      lazyLoad: 'progressive',
    },
    boxStyle: {},
    boxClassName: '',
  };

  it('to match snapshot', () => {
    const { baseElement } = render(<CarouselWrapper {...defaultProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('renders the correct number of slides', () => {
    const { getAllByTestId } = render(<CarouselWrapper {...defaultProps} />);
    const slides = getAllByTestId('slide');
    expect(slides.length).toBe(1);
  });

  it('renders the correct number of items per slide', () => {
    const { getAllByTestId } = render(<CarouselWrapper {...defaultProps} />);
    const items = getAllByTestId('item');
    expect(items.length).toBe(mockData.length);
  });

  it('renders the title correctly', () => {
    const { getByText } = render(<CarouselWrapper {...defaultProps} />);
    expect(getByText('Test Carousel')).toBeInTheDocument();
  });

  it('handles empty data gracefully', () => {
    const emptyProps = { ...defaultProps, data: [] };
    const { container } = render(<CarouselWrapper {...emptyProps} />);
    const slides = container.querySelectorAll('.slick-slide');
    expect(slides.length).toBe(0);
  });
});
