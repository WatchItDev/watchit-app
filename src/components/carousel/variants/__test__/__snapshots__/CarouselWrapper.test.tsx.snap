// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[COMPONENTS]: CarouselWrapper > to match snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-12jpqf3-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
          >
            Test Carousel
          </span>
        </div>
        <div
          class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="leftButton"
            tabindex="0"
            type="button"
          >
            <svg
              class="tabler-icon tabler-icon-chevron-left "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 6l-6 6l6 6"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="rightButton"
            tabindex="0"
            type="button"
          >
            <svg
              class="tabler-icon tabler-icon-chevron-right "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 6l6 6l-6 6"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-10jug1b-MuiCardContent-root"
      >
        <div
          class="MuiBox-root css-kdwx3d"
        >
          <div
            class="slick-slider slick-initialized"
          >
            <div
              class="slick-list"
              style="height: 0px;"
            >
              <div
                class="slick-track"
                style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-hidden="false"
                  class="slick-slide slick-active slick-current"
                  data-index="0"
                  style="outline: none; width: 0px;"
                  tabindex="-1"
                >
                  <div>
                    <div
                      data-testid="slide"
                      style="width: 100%; display: inline-block;"
                      tabindex="-1"
                    >
                      <div
                        data-testid="item"
                      >
                        Test Title 1
                      </div>
                      <div
                        data-testid="item"
                      >
                        Test Title 2
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
